torch/csrc/jit/passes/onnx/constant_fold.cpp

```diff
@@ -0,0 +1,187 @@
+#include <BaseTsd.h>
+typedef SSIZE_T ssize_t;
+#endif
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+using ParamMap = std::map<std::string, at::Tensor>;
+using ValueToParamPairMap = std::map<Value*, std::pair<std::string, at::Tensor>>;
+
+static ValueToParamPairMap buildValueToParamsMap(Block* b, const ParamMap& paramsDict) {
+  ValueToParamPairMap valsToParamsMap;
+  for(auto& input : b->inputs()) {
+      auto it = paramsDict.find(input->uniqueName());
+      if (it != paramsDict.end()) {
+          valsToParamsMap[input] = *it;
+      }
+  }
+  return valsToParamsMap;
+}
+
+static void buildParamsMapFromValueToParamsMap(
+    const ValueToParamPairMap& valsToParamsMap, ParamMap& paramsDict) {
+  paramsDict.clear();
+  for(auto& nameTensorParamPair : valsToParamsMap) {
+    paramsDict.insert(nameTensorParamPair.second);
+  }
+}
+
+static void eraseUnusedNodeOutputs(Node* node) {
+  std::vector<std::string> removedOutputNames;
+  for (size_t i_1 = node->outputs().size(); i_1 > 0; --i_1) {
+      size_t i = i_1 - 1;
+      if (!node->outputs().at(i)->hasUses()) {
+        node->eraseOutput(i);
+      }
+  }
+}
+
+static at::Tensor runTorchBackendForOnnx(const Node* node, std::vector<at::Tensor>& inputTensorValues) {
+  at::Tensor updated_val;
+  if (node->kind() == onnx::Slice) {
+    assert(inputTensorValues.size() == 1);
+    if ( !(node->hasAttributeS("axes") && node->hasAttributeS("starts") && node->hasAttributeS("ends")) ) {
+      throw std::runtime_error("Missing attribute(s) in onnx::Slice op.");
+    }
+    auto axesAttr = node->is(attr::axes);
+    auto startsAttr = node->is(attr::starts);
+    auto endsAttr = node->is(attr::ends);
+    if (axesAttr.size() != startsAttr.size() || axesAttr.size() != endsAttr.size()) {
+      throw std::runtime_error("onnx::Slice node attributues named, axes, starts, and ends, must be the same length.");
+    }
+    updated_val = inputTensorValues[0];
+    for (size_t i = 0; i < axesAttr.size(); ++i) {
+      updated_val = at::narrow(updated_val, axesAttr[i], startsAttr[i], endsAttr[i] - startsAttr[i]);
+    }
+  }  
+  else if (node->kind() == onnx::Concat) {
+    updated_val = at::cat(at::TensorList(inputTensorValues), node->i(attr::axis));
+  }
+  else if (node->kind() == onnx::Unsqueeze) {
+    assert(inputTensorValues.size() == 1);
+    if (!node->hasAttributeS("axes")) {
+      throw std::runtime_error("Missing attribute 'axes' in onnx::Unsqueeze op.");
+    }
+    updated_val = inputTensorValues[0];
+    for (auto axis: node->is(attr::axes)) {
+      updated_val = at::unsqueeze(updated_val, axis);
+    }
+  }
+  else if (node->kind() == onnx::Transpose) {
+    assert(inputTensorValues.size() == 1);
+    if (!node->hasAttributeS("perm")) {
+      throw std::runtime_error("Missing attribute 'perm' in onnx::Transpose op.");
+    }
+    updated_val = inputTensorValues[0].permute(node->is(attr::perm));
+  }
+  else {
+    updated_val = at::empty({0});
```

Comment:
I think it would be clearer for you to return a c10::optional<at::Tensor> from this function, and return `c10::nullopt` in the case that `node` is not a supported op for constant folding.

Reply:
Done.